tests/phpunit/tests/post.php

```diff
@@ -161,6 +190,97 @@ public function test_wp_count_posts() {
 		$this->assertEquals( new stdClass(), $count );
 	}
 
+	/**
+	 * Ensure `wp_count_posts()` in 'readable' context excludes private posts
+	 * authored by other users when the current user lacks the capability to
+	 * read private posts.
+	 *
+	 * @ticket 61097
+	 * @covers ::wp_count_posts
+	 */
+	public function test_wp_count_posts_readable_excludes_unreadable_private_posts() {
+		$post_type = rand_str( 20 );
+		register_post_type( $post_type );
+
+		$admin_user_id = self::$user_ids['administrator'];
+
+		self::factory()->post->create_many(
+			5,
+			array(
+				'post_type'   => $post_type,
+				'post_status' => 'publish',
+				'post_author' => $admin_user_id,
+			)
+		);
+
+		self::factory()->post->create_many(
+			3,
+			array(
+				'post_type'   => $post_type,
+				'post_status' => 'private',
+				'post_author' => $admin_user_id,
+			)
+		);
+
+		$current_user_id = self::$user_ids['author'];
+		wp_set_current_user( $current_user_id );
+
+		$count = wp_count_posts( $post_type, 'readable' );
+		$this->assertEquals( 5, $count->publish );
+		_unregister_post_type( $post_type );
+	}
+
+	/**
+	 * Tests the wp_count_posts_query filter
+	 *
+	 * @ticket 61097
+	 * @covers ::wp_count_posts
+	 */
+	public function test_wp_count_posts_query_filter() {
+		$post_type = rand_str( 20 );
+		register_post_type( $post_type );
+
+		$parent_post_id = self::factory()->post->create( array( 'post_type' => $post_type ) );
+
+		self::factory()->post->create_many(
+			2,
+			array(
+				'post_type' => $post_type,
+			)
+		);
+
+		self::factory()->post->create_many(
+			2,
+			array(
+				'post_type'   => $post_type,
+				'post_parent' => $parent_post_id,
+			)
+		);
+
+		$filter_query = function ( $query, $post_type ) use ( $parent_post_id ) {
+			global $wpdb;
+			return $wpdb->prepare(
+				"
+				SELECT post_status, COUNT(*) as num_posts 
+				FROM {$wpdb->posts} 
+				WHERE post_parent = %d AND post_type = %s
+				",
+				$parent_post_id,
+				$post_type
+			);
+		};
+
+		add_filter( 'wp_count_posts_query', $filter_query, 10, 2 );
+		$count = wp_count_posts( $post_type );
+		remove_filter( 'wp_count_posts_query', $filter_query, 10 );
+		$this->assertEquals( 2, $count->publish );
```

Comment:
How about this, as it adds the assertions that the expected args were passed into the filter callback:

```suggestion
		$filter_args  = null;
		$filter_query = function ( $query, $post_type ) use ( $parent_post_id, &$filter_args ) {
			$filter_args = func_get_args();
			global $wpdb;
			return $wpdb->prepare(
				"
				SELECT post_status, COUNT(*) as num_posts
				FROM {$wpdb->posts}
				WHERE post_parent = %d AND post_type = %s
				",
				$parent_post_id,
				$post_type
			);
		};

		add_filter( 'wp_count_posts_query', $filter_query, 10, PHP_INT_MAX );
		$count = wp_count_posts( $post_type );
		$this->assertEquals( 2, $count->publish );
		$this->assertIsArray( $filter_args );
		$this->assertCount( 2, $filter_args );
		$this->assertIsString( $filter_args[0] );
		$this->assertStringContainsString( 'SELECT post_status', $filter_args[0] );
		$this->assertSame( $post_type, $filter_args[1] );
```

Reply:
Good idea. But I would trim the asserts to the last two, as they implicitely hold thje rest (2 array arguments, and the string, is checked against a string).

Reply:
You could, but then when the args aren't passed the test will have an error rather than a failure.